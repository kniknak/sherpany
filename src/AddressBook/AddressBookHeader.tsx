import React from "react";
import { RootState } from "../App/store";
import Typography from "@material-ui/core/Typography";
import Toolbar from "@material-ui/core/Toolbar";
import AppBar from "@material-ui/core/AppBar";
import MuiTextField from "@material-ui/core/TextField";
import withStyles from "@material-ui/core/styles/withStyles";
import { Status } from "./redux/types";
import { connect, MapStateToProps } from "react-redux";

const TextField = withStyles(theme => ({
    root: {
        width: 100,
        marginLeft: theme.spacing(2),
    },
}))(MuiTextField);

interface StateProps {
    status: Status
}

interface OwnProps {
    status: Status
    search: string
    setSearch: (search: string) => void
}

export type AddressBookHeaderProps = StateProps & OwnProps

export class AddressBookHeader extends React.PureComponent<AddressBookHeaderProps> {
    private onChange: React.ChangeEventHandler<HTMLInputElement> = event => this.props.setSearch(event.target.value)

    render() {
        const { status, search } = this.props

        return (
            <AppBar position="sticky"
                color="default">
                <Toolbar style={{ alignItems: "baseline" }}>
                    <Typography variant="body1">
                        {Status[status]}
                    </Typography>
                    <Typography
                        variant="body1"
                        style={{ flex: 1 }}
                        align="right"
                    >
                        Search:
                    </Typography>
                    <TextField
                        label="Type text"
                        value={search}
                        onChange={this.onChange}
                        margin="normal"
                        variant="outlined"
                        size="small"
                    />
                </Toolbar>
            </AppBar>
        );
    }
}


const mapStateToProps: MapStateToProps<StateProps, {}, Pick<RootState, "addressBook">> = state => ({
    status: state.addressBook.status,
});

export const AddressBookHeaderConnected = connect(mapStateToProps)(AddressBookHeader);

