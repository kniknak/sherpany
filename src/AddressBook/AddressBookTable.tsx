import React from "react";
import { RootState } from "../App/store";
import { connect, MapStateToProps } from "react-redux";
import Table from "@material-ui/core/Table";
import TableHead from "@material-ui/core/TableHead";
import TableRow from "@material-ui/core/TableRow";
import TableCell from "@material-ui/core/TableCell";
import TableBody from "@material-ui/core/TableBody";
import { AddressBookPerson } from "./Person/AddressBookPerson";
import throttle from "lodash.throttle";
import { fetchAddressBook } from "./redux/fetchAddressBook";
import { FetchAddressBookUser, Status } from "./redux/types";
import { Preloader } from "../UI/Preloader";
import { AddressBookPersonModal } from "./Person/AddressBookPersonModal";

const isPersonShown = (person: FetchAddressBookUser, search: string) => (search === "") || (
    [person.name.first, person.name.last]
        .join(" ")
        .toLowerCase()
        .includes(search.toLowerCase())
);

interface StateProps {
    users: FetchAddressBookUser[][]
    status: Status
}

interface DispatchProps {
    fetchAddressBook: (page: number) => void
}

interface OwnProps {
    search: string
    scrollThreshold: number
    maxPage: number

}

export type AddressBookTableProps = StateProps & DispatchProps & OwnProps

interface State {
    page: number
    isOpen: boolean
    person?: FetchAddressBookUser
}

export class AddressBookTable extends React.PureComponent<AddressBookTableProps, State> {
    private throttleInterval = 500;

    readonly state: State = {
        page: 1,
        isOpen: false,
    };

    componentDidMount() {
        window.addEventListener("scroll", this.onScroll);
        this.fetchResults(this.state.page);
    }

    componentWillUnmount() {
        window.removeEventListener("scroll", this.onScroll);
    }

    private onScroll = () => {
        if (this.props.search !== "") {
            return;
        }

        if (window.innerHeight + document.documentElement.scrollTop >= document.documentElement.offsetHeight - this.props.scrollThreshold) {
            if (this.props.status !== Status.Pending) {
                const page = Math.min(this.props.maxPage, this.state.page + 1);
                this.setState({ page });
                this.fetchResults(page);
            }
        }

    };

    private fetchResults = throttle((page: number) => this.props.fetchAddressBook(page), this.throttleInterval);

    private onOpen = (person: FetchAddressBookUser) => this.setState({
        isOpen: true,
        person,
    });

    private onClose = () => this.setState({
        isOpen: false,
        person: undefined,
    });

    private renderPage = (slice: FetchAddressBookUser[], page: number) => (
        <React.Fragment key={page}>
            {slice.map(this.renderPerson)}
        </React.Fragment>
    );

    private renderPerson = (person: FetchAddressBookUser) => (
        <AddressBookPerson
            key={person.id.value}
            person={person}
            isShown={isPersonShown(person, this.props.search)}
            onOpen={this.onOpen}
        />
    );

    render() {
        const { users, maxPage, status } = this.props;
        const { page, isOpen, person } = this.state;

        return (
            <>
                <Table>
                    <TableHead>
                        <TableRow>
                            <TableCell>picture.thumbnail</TableCell>
                            <TableCell>name.first</TableCell>
                            <TableCell>name.last</TableCell>
                            <TableCell>name.username</TableCell>
                            <TableCell>email</TableCell>
                            <TableCell/>
                        </TableRow>
                    </TableHead>
                    <TableBody>
                        {users.map(this.renderPage)}
                    </TableBody>
                </Table>
                {status === Status.Pending ? (
                    <Preloader>loading...</Preloader>
                ) : (
                    (page === maxPage) && (
                        <Preloader>end of users catalog</Preloader>
                    )
                )}
                <AddressBookPersonModal
                    person={person}
                    isOpen={isOpen}
                    onClose={this.onClose}
                />
            </>
        );
    }
}

const mapStateToProps: MapStateToProps<StateProps, OwnProps, Pick<RootState, "addressBook">> = state => ({
    users: state.addressBook.users,
    status: state.addressBook.status,
});

const mapDispatchToProps: DispatchProps = {
    fetchAddressBook,
};
export const AddressBookTableConnected = connect(mapStateToProps, mapDispatchToProps)(AddressBookTable);
