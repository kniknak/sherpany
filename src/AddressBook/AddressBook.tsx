import React, { useState, memo } from "react";
import { AddressBookTableConnected } from "./AddressBookTable";
import { AddressBookHeaderConnected } from "./AddressBookHeader";

const scrollThreshold = 500;
const maxPage = 20;

export const AddressBook: React.FC = memo(() => {
    const [search, setSearch] = useState("");

    return (
        <>
            <AddressBookHeaderConnected
                search={search}
                setSearch={setSearch}
            />
            <AddressBookTableConnected
                search={search}
                scrollThreshold={scrollThreshold}
                maxPage={maxPage}
            />
        </>
    );
});

export default AddressBook;