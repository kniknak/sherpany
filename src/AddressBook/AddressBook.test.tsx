import React from "react";
import { shallow } from "enzyme";
import { AddressBook } from "./AddressBook";

describe("AddressBook", () => {
    describe("AddressBook", () => {
        it("renders", () => {
            const subject = shallow(<AddressBook/>);

            expect(subject).toMatchSnapshot()
        });
    });
});