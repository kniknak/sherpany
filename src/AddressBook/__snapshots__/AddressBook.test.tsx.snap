// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddressBook AddressBook renders 1`] = `
<Fragment>
  <Connect(AddressBookHeader)
    search=""
    setSearch={[Function]}
  />
  <Connect(AddressBookTable)
    maxPage={20}
    scrollThreshold={500}
    search=""
  />
</Fragment>
`;
